Step3 : 자동차 경주

Mo-bile opened the pull request:
안녕하세요? 모재영입니다!
step3 자동차 경주 PR 드립니다 👍 

## 첫 TDD 후기
제안주신 방식대로 TDD로 모두 구현을 시도했습니다.
테스트 부터 먼저 만드는 것이 어려웠습니다. 본 과제 수행에 투여한 시간의 절반이 테스트 만들기 였습니다.
어렵지만 막상 테스트를 짜면서 놓쳤던 기능 명세가 한번 더 되고 막연하게 생각했던 구현 방향이 잡히게 되었습니다.
흥미로운 경험이었습니다. (하지만 어렵네요 하하하)


## 질문
1. 테스트 코드를 먼저짜면서 이런 방식이 맞을까요?
테스트 코드만 짜는데 2시간 30분이 걸렸는데요. 생각보다 오래걸려서 놀랬습니다.
그리고 막상 테스트 코드를 짜면서 'model 을 정의해야겠다!', 'util 로 뽑아내야겠다' 이런생각이 들었습니다 그러면서 뒤늦게 기능 개발이 추가 된 느낌입니다. readme 작성도 뒤늦게 추가되기도 하고요...

2. 테스트 코드 작성 시 접근제어자가 private 인경우에는 테스트를 못한다고 보는게 맞을까요?
그래도 무언가 로직단위인데 테스트를 못하니 아쉬워서요... 상위 메서드를 호출해서 우회적으로 테스트 하는 느낌이었습니다

4. 프로덕션 코드 작성 중이라도 놓친 테스트가 생각나거나 하면 얼마든지 뒤늦게 추가해도 되는것 맞을까요?
이것은 주요 TDD 싸이클 중 어디로 보면 좋을까요? 자바지기님의 경우에는 어떻게 하실까요?

5. 처음에 TDD 싸이클 중 red 에서는 컴파일 에러나 문법 틀린채로 넘어가는것 맞을까요?
혹은 껍대기만 있는 메서드라도 프로덕션 코드에 만들기 까지 필요할까요?

## 프로젝트 구조 설명
### 구조도
```
RacingGame
├── Business
│   └── Process
│
├── model
│   └── Car
│
├── util
│   └── RandomUtil
│
├── GameApplication
├── InputView
└── ResultView

```
### 구조 설명
- Business : 게임의 주요 흐름과 주요 로직을 담당합니다. 
내부에 Process 클래스가 존재하여 인스턴스로 GameApplication 에서 실행합니다.

- model : 도메인 모델인 Car 에 대한 정의가 있습니다
car 모델에는 현재위치관련 상태를 가진 필드와 초기화, 이동, 현재위치를 알려주는 메서드가 있습니다

- util : 랜덤 정수 값 출력 관련 RandomUtil 클래스가 있습니다.

## 프로그램 설명
### 실행방법
- 기본적인 프로그램 실행은 `GameApplication`클래스의 main 메서드에서 실행을 합니다.
이 경우 InputView 를 통한 레이싱 게임 참여 자동차 와 게임 횟수 관련 값을 입력 받을 수있습니다.
(테스트 코드로 실행시에는 값을 입력 할 수없습니다)

- 그외 테스트 코드 실행은 Process 클래스의 run() 메서드를 곧바로 호출하여 테스트를 진행합니다.

## 주요 로직 설명
프로그램의 큰 흐름은 `Process`클래스의 run() 메서드에서 실행됩니다.
1. init()
참가하는 자동차 생성과, 게임 이동횟수를 입력받는다.
2. 게임 이동횟수 만큼 반복문을 돌면서 `nextMove()` 메서드를 호출한다


## 나가며
긴 글 읽어주셔서 감사드립니다
금번 리뷰도 잘 부탁드리겠습니다 👍

Commit:
add dependency junit-jupiter-params

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Mo-bile committed:
Merge branch 'next-step:mo-bile' into mo-bile

Commit:
docs : 구현할 기능 목록 정리하여
입력, 출력, 진행으로 구분

Commit:
doc : 누락된 기능목록 추가

Commit:
doc : 누락된 기능목록 추가 (자동차 자료형)

Commit:
test : 실패하는 GameProcessTest, CarTest, RandomUtilTest 추가

Commit:
feat : 랜덤 유틸 구현 및 리팩터링 완료

Commit:
test : 랜덤 범위 초과 false 조건 추가 완료

Commit:
feat : car model 개발

Commit:
test : 변경된 car model 기능에 따라 test 변경완료

Commit:
test : 불명확한 기능 및 실행 흐름 다시 재정의하여 테스트 완료

Commit:
refactor : 프로젝트 구조 재정의 및 클래스명 변경하여 유지보수성 및 가독성 향상

Commit:
test : 테스트 수정하여 run() 에 대한 예외를 받을 수있게 변경 완료

Commit:
test : 테스트에 예외처리 에러 메시지 재정의

Commit:
feat : 참가시킬 자동차 수 예외처리 개발

Commit:
feat : 이동횟수 테스트 통과 및 리패터링 완료

Commit:
feat : 참가하는 자동차 기본 세팅 및 리팩터링 완료

Commit:
feat : 게임 이동횟수 기능 개발 및 리팩터링 완료

Commit:
feat : 자동차 이동 기능 개발 및 리팩터링 완료

Commit:
feat : 여러 이동, 여러 자동차 기능 및 리팩터링 완료

Commit:
feat : 기본 입출력 개발

Commit:
fix : car 초기화 변수 0에서 1로 수정

Commit:
refactor
1. 출력문 위치 조절하여 출력문 출력 예상가능하게 함

2. 각 출력문 메서드화

javajigi commented on src/main/java/racingGame/Business/Process.java:
```diff
@@ -0,0 +1,62 @@
+import racingGame.model.Car;
+import racingGame.util.RandomUtil;
+
+public class Process {
```
구현한 코드를 살펴보니 핵심 비즈니스 로직(예: 자동차 이동)과 입출력 로직(System.in, System.out)이 하나의 클래스나 메서드에 섞여 있는 것 같습니다. 

질문:
1. 현재 코드가 가진 핵심 로직을 (키보드 입력이나 화면 출력 없이) 순수한 기능으로만 분리하려면 어떻게 수정해야 할까요?
2. 두 로직을 분리하지 않고 한 곳에 두었을 때, 테스트 용이성이나 코드를 다른 환경(예: 웹, 모바일)에서 재사용하는 측면에서 어떤 문제점이 발생할까요?

힌트:
테스트하기 어려운 코드(UI, DB 접근)와 테스트하기 쉬운 코드(순수 도메인 로직)를 분리하여 테스트하기 쉬운 코드를 TDD로 구현하는 것을 목표로 합니다.
이 연습은 웹 프로그래밍에서 DB에 종속된 코드와 DB에 종속되지 않은 코드를 분리하는 것과 같은 원리입니다. 핵심 비즈니스 로직을 UI나 DB와 같은 인프라 영역으로부터 독립시키세요.

이 부분만 확실히 이해하고 설계한다면 훨씬 단위 테스트 진행하기도 수월해지고 설계 역량도 향상될 겁니다. 힘들 수 있겠지만 한번 도전해 보면 어떨까요?

javajigi commented on src/main/java/racingGame/Business/Process.java:
```diff
@@ -0,0 +1,62 @@
+public class Process {
+    
+    public static final int CAR_FORWARD_CRITERIA = 3;
+    private final ArrayList<Car> cars = new ArrayList<>();
```
```suggestion
    private final List<Car> cars = new ArrayList<>();
```

javajigi commented on src/main/java/racingGame/Business/Process.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+    
+    public ArrayList<Car> joinCars() {
```
```suggestion
    public List<Car> joinCars() {
```
내부 구현체가 LinkedList와 같은 구현체로 변경하더라도 외부에 변경이 발생하지 않음. 따라서 인터페이스가 있다면 인터페이스를 외부에 노출할 것을 추천

Mo-bile replied:
DIP 원칙을 잊고있었네요 짚어주셔서 감사드립니다!

javajigi commented on src/main/java/racingGame/model/Car.java:
```diff
@@ -0,0 +1,21 @@
+    public Car() {
+    }
+    
+    public void init() {
+        this.location = 1;
+    }
```
```suggestion
    public Car() {
        this.location = 1;
    }
```

생성할 때 값을 초기화할 것을 추천해요.
init() 메서드를 분리함으로 인해 Car 생성할 때마다 init() 메서드를 호출해야하고, init() 메서드를 호출하지 않을 경우 값이 초기화되지 않는 문제가 생기지 않을까요?

javajigi commented on src/main/java/racingGame/InputView.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.Scanner;
+
+public class InputView {
```
InputView와 ResultView를 클래스 메서드(static)으로 구현 👍

javajigi commented on src/test/java/racingGame/Business/ProcessTest.java:
```diff
@@ -0,0 +1,71 @@
+import org.junit.jupiter.api.Test;
+import racingGame.model.Car;
+
+class ProcessTest {
```
앞에도 피드백했지만 Process가 핵심 비지니스 로직과 UI 연결 로직도 포함하고 있어 복잡도도 높고 단위 테스트하기도 힘들어 집니다.
이 테스트 View와 연결된 부분도 있어 진정한 단위 테스트로 보기 힘들 것 같아요.
Process에서 핵심 비지니스 로직(model)과 UI 로직을 분리하고 핵심 비지니스 로직(model)에 대해서만 단위 테스트 추가해볼 것을 추천합니다.

javajigi reviewed (changes requested):
pr 본문 내용을 보니 TDD를 도전하기 위한 여정이 느껴지네요. 💯 
그런 힘든 여정이 있었기 때문에 그에 따른 다양한 질문이 발생한 것 같아요. 굿

답변을 하기 전에 피드백에도 남겼는데요.
이번 기회에 확실히 핵심 비지니스 로직과 UI 로직을 분리하는 연습을 꼭꼭꼭 해보면 좋겠습니다.
그렇게 할 때 TDD가 가능해 집니다.

Q1. 테스트 코드를 먼저짜면서 이런 방식이 맞을까요?
테스트 코드만 짜는데 2시간 30분이 걸렸는데요. 생각보다 오래걸려서 놀랬습니다.
그리고 막상 테스트 코드를 짜면서 'model 을 정의해야겠다!', 'util 로 뽑아내야겠다' 이런생각이 들었습니다 그러면서 뒤늦게 기능 개발이 추가 된 느낌입니다. readme 작성도 뒤늦게 추가되기도 하고요...
A1. TDD가 처음이라 지금 단계는 이런 혼란스러움을 느끼는 것이 맞아요.
지금은 오래 걸리지만 익숙해지면 점점 더 짧아지니 넘 걱정하지 마세요.
아마도 어떻게 시작해야할 것인지 몰라서 그런 건데요.
답답할 수 있지만 그 단계를 넘으면 다른 세상과 접근 방식을 느낄 수 있을 겁니다.

Q2. 테스트 코드 작성 시 접근제어자가 private 인경우에는 테스트를 못한다고 보는게 맞을까요?
그래도 무언가 로직단위인데 테스트를 못하니 아쉬워서요... 상위 메서드를 호출해서 우회적으로 테스트 하는 느낌이었습니다
A2. private 메서드를 테스트하고 싶은 건 대부분의 경우 설계를 잘못했기 때문일 수 있어요. 이럴 때 설계를 개선할 부분이 있는 것이 아닐까라는 힌트로 생각해 보면 좋겠습니다. 추후 강의 중에도 다룰 겁니다.

Q3. 프로덕션 코드 작성 중이라도 놓친 테스트가 생각나거나 하면 얼마든지 뒤늦게 추가해도 되는것 맞을까요?
이것은 주요 TDD 싸이클 중 어디로 보면 좋을까요? 자바지기님의 경우에는 어떻게 하실까요?
A3. 당연하죠. 우리가 신이 아닌 이상 처음부터 완벽하게 설계할 수는 없잖아요. 저도 똑같아요. 구현하면서 생각날 때마다 todo list에 추가해 나가면서 기능 개발하면 됩니다. 이 접근 방식이 당연한 접근 방식이니 초반에 완벽한 설계를 하기 위해 너무 많은 시간을 쓰지 않기를 추천해요.

Q4. 처음에 TDD 싸이클 중 red 에서는 컴파일 에러나 문법 틀린채로 넘어가는것 맞을까요?
혹은 껍대기만 있는 메서드라도 프로덕션 코드에 만들기 까지 필요할까요?
A4. 첫 번째 test case 추가할 때 일시적으로 컴파일 에러가 발생하는 것 맞아요. 단 red 단계에서 컴파일 에러 해결하고, 실패하는 테스트를 보기 위해 껍대기만 있는 메서드라도 프로덕션 코드에 만드는 것이 맞아요. 그래야 실패하는 테스트를 실행해 볼 수 있으니까요.

Commit:
refactor : Car doamin에 대한 Init() 메서드를 제거하고 생성자에 추가하여 init() 메서드 호출 실수 방지

Commit:
refactor : DIP 원칙에 따라 추상화에 의존해야 하기 때문에 하위수준인 'ArrayList' 종속 제거

Commit:
test : RacingGameRule 모델 추가하여 레이싱 게임 자체의 rule 에 대한 역할을 분리 및 테스트

Commit:
refactor : RacingGameRules 의 인스턴스 생성 방식을 생성자에서 정적 팩터리 메서드 패턴으로 변경하여, 외부 입력에 대한 검증 및 게이트웨이 역할 표현 가능

Commit:
feat : Rules 모델에 대한 행동 테스트 추가하여 현재 게임 rule 준수 여부 체크 가능

Commit:
refactro : 입력순서에 따라 인자 순서 변경하여 전체 process 내 메서드의 인자와 일치시킴

Commit:
refactor :
1. init() 에 자동차 생성과 게임 이동횟수 제거하여 기본 게임 Rule 세팅에 집중하게 변경 완료
2. run() 에 participateCar() 라는 메서드를 분리하여 입력한 만큼의 자동차 생성방식으로 변경하여 자동차 생성 흐름이 있음을 볼수있게 변경완료
2. cars 를 joinCars로 변경하여 list에대한 목적 명확화

Commit:
feat : 레이싱 기록을 위한 model 을 만들어서 현재 게임 정도 체크와 레이스 기록을 체크 할 수있게 함

Commit:
feat : gameRecord() 인자를 car 단일에서 List로 변경하여 한번에 기록할 수있게끔 변경

Commit:
refactor :
1. progressRecords 에다가 게임을 기록하여 게임 진행현황과 게임 과거 기록을 저장
2. run() 메서드에 게임 결과를 리턴하여 출력 책임을 제거

Commit:
refactor
1. 게임 결과 출력을 기본 타입으로 변경하여 ResultView의 static 메서드인 printAllResult 가 비즈니스 로직의 도메인 및 상태를 모르게 변경
2. 이를 위해 ProgressRecord 에 Car 도메인을 모르는 Interger 형 단순 기록 자로형 추가 및 변환 출력 추가

Commit:
refactor : 랜덤 출력 시 이동 기준을 Rules 도메인으로 이동시켜 메인 Process에서는 게임 흐름에만 역할 집중에게 변경

Commit:
refactor
1. Process 에 주요 메서드 위치 변경
2. ProcessTest 에 일반테스트에서ParameterizedTest 로 변경
3. RulesTest 에 잘못된 매개변수 순서 변경

Commit:
refactor : isProceedByRandom() 메서드 조건 단순화

Mo-bile commented:
안녕하세요? 👍 
이전에 드린 PR에 커멘트 남긴 뒤 리뷰 요청 드리려고 했습니다.
그런데 실수로 본 [PR](https://github.com/next-step/java-racingcar/pull/6189) 과 관련된 origin branch 를 삭제를 해서 PR 이 closed 가 됐네요... 😂 이 부분 양해부탁드리겠습니다. 저도 당황스럽네요... close 된 PR re-open 하는 방식을 시도했지만 'The step3 branch was force-pushed or recreated.' 라면서 불가능 하다고 합니다... 최종 merge 를 위해 다시 PR을 열어야할까요?

두번째 PR은 이전에 요청주신 '핵심 비지니스 로직과 UI 로직을 분리하는 연습' 에 집중하였습니다.
특히 출력부분을 비즈니스 로직에서 분리하는데 새로운 model 이 필요하고 클래스 분리까지 고민하게 되었습니다.

## 질문
### 1. 역할과 책임에 집중하여 클래스를 분리하려고 하는데 이부분이 쉽지 않았습니다. 어떤 기준으로 나누는게 좋을까요?
클래스 분리에 책임단위라는 느낌이 어려워서 개인적으로 '맥락(context)' 가 달리한다면 나눈다는 기준을 두었습니다.
왜나하면 각각의 역할과 책임은 맥락이 다르기 때문입니다. 어떻게 생각하실까요?

그럼에도 불구하고 맥락에 따라 나누려고하는데 클래스를 보는 맥락을 어떤관점으로 보느냐에 따라 달리 볼수있겠구나 싶어습니다.
예로들어서 현실의 세계에서도 `경찰이 보는 사람`, `의사가 보는 사람`, `은행원이 보는 사람` 모두 다른것 처럼요
개인적으로 알아보니 개발에서는 domain, actor, 기능 관점으로 나눈다고 하는데 혹시 어떤 기준으로 나누시는지요?

### 2. 과도한 추상화를 멈추는 기준이 있을까요?
무언가 클래스 분리, 메서드 분리를 하려고하는데 욕심이 과하다 보니 과도한 추상화를 하고있다고 생각이 들어서 몇번 멈추고 되돌렸습니다.
롤백하면서 내다보낸 시간이 아쉬웠습니다. 혹시 어떤기준으로 고민하면 좋을까요?

주요 예)
`Process` 클래스의 `joinCars`,`progressRecords`와 필드가 List 형인데 이 부분을 한번 더 추상화해서 사용할까 생각했습니다.
그런데 과도한 캡슐화로 이용이 어려워지고 이 부분에 대한 행동정의가 어려워서 포기했습니다.
포기하면서 아쉬운 점은 이 필드에 대한 반복문이나 여러 처리가 결국에 `Process` 에서 이루어 지는 것이 만족스럽지 못했습니다. process에 이 기능이 있는것이 바람직하지 못하다고 생각하기 때문입니다.

## 주요 의도
이번에 추가 변경된 각 클래스나 코드 별 어떤 의도인지 말씀드리겠습니다

### `Process`
- 앞서 언급드린 것 처럼 `joinCars`,`progressRecords` 필드를 한번 더 model 로 감싸려고 했는데 과도한 추상화 및 행동 지정이 어려워서 포기했습니다.
- `showResult()` 는 출력을 위한 데이터로 변환화는 것인데요 특히 progressRecords가 List형이다 보니 여기서 어쩔수없이 처리했습니다.
  - Process 에는 적합하지 않다고 들어서 고민이 지금도 있네요 이것을 다른곳으로 옮겨야하나 싶습니다

### `ProgressRecord`
- 가장 큰 목적은 각 횟수마다 게임 기록입니다.
- `simpleResult`은 ResultView 에 전달하는 목적입니다. ResultView 에 게임 진행관련 상태를 전달하고 싶지않아서 리팩터링하면서 추가된 메서드입니다.
  - 그러면서 `simpleResult`가 이미 이 있는데 `joinCarsRecord`의 인자에 Car가 있는채로 Car를 의존해야할까 싶었습니다.(즉 `ProgressRecord` 가 'Car'를 의존해야하는가?) 그럼에도 불구하고`ProgressRecord` 에 model class의 책임은 Car에 대한 위치 기록이 필요하다고 생각해서 유지했습니다.
  - Car의 위치를 단순 정수형으로 변환을 위해 `ProgressRecord`에 `makeSimpleGameResult()` 변환용 메서드를 따로 만들었습니다

### `Rules`
- 인스턴스 생성을 생성자 대신에 `of`를 이용했습니다. 다른 수강생 분들의 리뷰를 보면서 `of` 사용에 대한 고민을 요청주셔서 사용해도 될까 고민 했습니다.
  - 고민 결과 Rules 의 인스턴스 생성은 외부 I/O(API 입력, DB 조회)로 부터 값을 받아서 생성하기 때문에 필요하다고 결정했습니다.
  - 즉 '외부 입력에 대한 검증' 및 '비즈니스 로직으로의 게이트웨이 역할' 표현 가능에 목적을 두었습니다.
- 게임 규칙관련임 처음에는 입력받은 두개 필드(`cars` ,`moves`) 추가하고 , 자동차 전진 기준인 `CAR_FORWARD_CRITERIA` 과 `isProceedByRandom(int randomNum)` 를 추가하였습니다. Random 관련 Util을 모르게 하고자 외부에서 int randomNum 인자로 받아서 작동하게만 했습니다


### `ProcessTest`
- 게임 비즈니스 로직을 테스트 코드 짜면서 방식에 고믾이 많았습니다. 각 게임 입력 상태값 별 결과 출력되는 값이나 상태를 조회하는 것에 어려움이 있었습니다.
  - 그러다보니 탄생한 메서드가 `Process.joinCars()` 입니다. 프로덕션 코드에는 이용되지 않고 테스트 코드를 위해 탄생한 메서드인데 이게 최선인가 고민이 많았습니다.
  - 만약 이 방식에 지양이 권장된다면 어떤 방식으로 접근하면 좋을까요?

## 나가며
요청 주신바대로 적용하면 금방하겠지 하지만, 어제 보다 시간을 쓴 것 같고, 더 많은 고민이 필요한 시간같았습니다.
그러면서 메서드 분리, 클래스 분리의 목적이 무엇인지 계속 고민하게 된 시간같습니다.

좋은 기회 주셔서 감사드립니다 👍

Mo-bile commented:
안녕하세요? 👍 
이전에 드린 PR에 커멘트 남긴 뒤 리뷰 요청 드리려고 했습니다.
그런데 실수로 본 [PR](https://github.com/next-step/java-racingcar/pull/6189) 과 관련된 origin branch 를 삭제를 해서 PR 이 closed 가 됐네요... 😂 이 부분 양해부탁드리겠습니다. 저도 당황스럽네요... close 된 PR re-open 하는 방식을 시도했지만 'The step3 branch was force-pushed or recreated.' 라면서 불가능 하다고 합니다... 최종 merge 를 위해 다시 PR을 열어야할까요?

두번째 PR은 이전에 요청주신 '핵심 비지니스 로직과 UI 로직을 분리하는 연습' 에 집중하였습니다.
특히 출력부분을 비즈니스 로직에서 분리하는데 새로운 model 이 필요하고 클래스 분리까지 고민하게 되었습니다.

## 질문
### 1. 역할과 책임에 집중하여 클래스를 분리하려고 하는데 이부분이 쉽지 않았습니다. 어떤 기준으로 나누는게 좋을까요?
클래스 분리에 책임단위라는 느낌이 어려워서 개인적으로 '맥락(context)' 가 달리한다면 나눈다는 기준을 두었습니다.
왜나하면 각각의 역할과 책임은 맥락이 다르기 때문입니다. 어떻게 생각하실까요?

그럼에도 불구하고 맥락에 따라 나누려고하는데 클래스를 보는 맥락을 어떤관점으로 보느냐에 따라 달리 볼수있겠구나 싶어습니다.
예로들어서 현실의 세계에서도 `경찰이 보는 사람`, `의사가 보는 사람`, `은행원이 보는 사람` 모두 다른것 처럼요
개인적으로 알아보니 개발에서는 domain, actor, 기능 관점으로 나눈다고 하는데 혹시 어떤 기준으로 나누시는지요?

### 2. 과도한 추상화를 멈추는 기준이 있을까요?
무언가 클래스 분리, 메서드 분리를 하려고하는데 욕심이 과하다 보니 과도한 추상화를 하고있다고 생각이 들어서 몇번 멈추고 되돌렸습니다.
롤백하면서 내다보낸 시간이 아쉬웠습니다. 혹시 어떤기준으로 고민하면 좋을까요?

주요 예)
`Process` 클래스의 `joinCars`,`progressRecords`와 필드가 List 형인데 이 부분을 한번 더 추상화해서 사용할까 생각했습니다.
그런데 과도한 캡슐화로 이용이 어려워지고 이 부분에 대한 행동정의가 어려워서 포기했습니다.
포기하면서 아쉬운 점은 이 필드에 대한 반복문이나 여러 처리가 결국에 `Process` 에서 이루어 지는 것이 만족스럽지 못했습니다. process에 이 기능이 있는것이 바람직하지 못하다고 생각하기 때문입니다.

## 주요 의도
이번에 추가 변경된 각 클래스나 코드 별 어떤 의도인지 말씀드리겠습니다

### `Process`
- 앞서 언급드린 것 처럼 `joinCars`,`progressRecords` 필드를 한번 더 model 로 감싸려고 했는데 과도한 추상화 및 행동 지정이 어려워서 포기했습니다.
- `showResult()` 는 출력을 위한 데이터로 변환화는 것인데요 특히 progressRecords가 List형이다 보니 여기서 어쩔수없이 처리했습니다.
  - Process 에는 적합하지 않다고 들어서 고민이 지금도 있네요 이것을 다른곳으로 옮겨야하나 싶습니다

### `ProgressRecord`
- 가장 큰 목적은 각 횟수마다 게임 기록입니다.
- `simpleResult`은 ResultView 에 전달하는 목적입니다. ResultView 에 게임 진행관련 상태를 전달하고 싶지않아서 리팩터링하면서 추가된 메서드입니다.
  - 그러면서 `simpleResult`가 이미 이 있는데 `joinCarsRecord`의 인자에 Car가 있는채로 Car를 의존해야할까 싶었습니다.(즉 `ProgressRecord` 가 'Car'를 의존해야하는가?) 그럼에도 불구하고`ProgressRecord` 에 model class의 책임은 Car에 대한 위치 기록이 필요하다고 생각해서 유지했습니다.
  - Car의 위치를 단순 정수형으로 변환을 위해 `ProgressRecord`에 `makeSimpleGameResult()` 변환용 메서드를 따로 만들었습니다

### `Rules`
- 인스턴스 생성을 생성자 대신에 `of`를 이용했습니다. 다른 수강생 분들의 리뷰를 보면서 `of` 사용에 대한 고민을 요청주셔서 사용해도 될까 고민 했습니다.
  - 고민 결과 Rules 의 인스턴스 생성은 외부 I/O(API 입력, DB 조회)로 부터 값을 받아서 생성하기 때문에 필요하다고 결정했습니다.
  - 즉 '외부 입력에 대한 검증' 및 '비즈니스 로직으로의 게이트웨이 역할' 표현 가능에 목적을 두었습니다.
- 게임 규칙관련임 처음에는 입력받은 두개 필드(`cars` ,`moves`) 추가하고 , 자동차 전진 기준인 `CAR_FORWARD_CRITERIA` 과 `isProceedByRandom(int randomNum)` 를 추가하였습니다. Random 관련 Util을 모르게 하고자 외부에서 int randomNum 인자로 받아서 작동하게만 했습니다


### `ProcessTest`
- 게임 비즈니스 로직을 테스트 코드 짜면서 방식에 고믾이 많았습니다. 각 게임 입력 상태값 별 결과 출력되는 값이나 상태를 조회하는 것에 어려움이 있었습니다.
  - 그러다보니 탄생한 메서드가 `Process.joinCars()` 입니다. 프로덕션 코드에는 이용되지 않고 테스트 코드를 위해 탄생한 메서드인데 이게 최선인가 고민이 많았습니다.
  - 만약 이 방식에 지양이 권장된다면 어떤 방식으로 접근하면 좋을까요?

## 나가며
요청 주신바대로 적용하면 금방하겠지 하지만, 어제 보다 시간을 쓴 것 같고, 더 많은 고민이 필요한 시간같았습니다.
그러면서 메서드 분리, 클래스 분리의 목적이 무엇인지 계속 고민하게 된 시간같습니다.

좋은 기회 주셔서 감사드립니다 👍
